vignettes/opt-cond-thread.Rmd

```diff
@@ -0,0 +1,81 @@
+---
+output: rmarkdown::html_vignette
+title: Conditional Threading Optimizer
+vignette: >
+  %\VignetteIndexEntry{Conditional Threading Optimizer}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r echo=FALSE, message=FALSE}
+library("rco")
+library("microbenchmark")
+library("ggplot2")
+autoplot.microbenchmark <- function(obj) {
+  levels(obj$expr) <- paste0("Expr_", seq_along(levels(obj$expr)))
+  microbenchmark:::autoplot.microbenchmark(obj)
+}
+speed_up <- function(obj) {
+  levels(obj$expr) <- paste0("Expr_", seq_along(levels(obj$expr)))
+  obj <- as.data.frame(obj)
+  summaries <- do.call(rbind, by(obj$time, obj$expr, summary))
+  res <- c()
+  for (i in seq_len(nrow(summaries) - 1) + 1) {
+    res <- rbind(res, summaries[1, ] / summaries[i, ])
+  }
+  rownames(res) <- levels(obj$expr)[-1]
+  return(res)
+}
+```
+
+# Conditional Threading Optimizer
+
+## Background
+
+Conditional Threading is an optimization that searches for separate blocks of `IF` statements but with exactly the same conditions. When several conditionals present the same boolean condition, merging their inside actions would avoid computing the condition several times. Also, in the cases where a second `IF` statement has been used with the exact negation of the condition of the first `IF`, the second `IF` could then be incorporated in an `ELSE` statement.
+
+## Example
+
+Consider the following example:
+
+```{r}
+code <- paste(
```

Comment:
When building the vignette I am getting that for this example the unoptimized code is faster.
Please provide an example which shows the speedup

Reply:
this is the full code to reproduce the vignette.
```
library("rco")
library("microbenchmark")
library("ggplot2")
autoplot.microbenchmark <- function(obj) {
  levels(obj$expr) <- paste0("Expr_", seq_along(levels(obj$expr)))
  microbenchmark:::autoplot.microbenchmark(obj)
}
speed_up <- function(obj) {
  levels(obj$expr) <- paste0("Expr_", seq_along(levels(obj$expr)))
  obj <- as.data.frame(obj)
  summaries <- do.call(rbind, by(obj$time, obj$expr, summary))
  res <- c()
  for (i in seq_len(nrow(summaries) - 1) + 1) {
    res <- rbind(res, summaries[1, ] / summaries[i, ])
  }
  rownames(res) <- levels(obj$expr)[-1]
  return(res)
}

code <- paste(
  "is_even <- function(a) {",
  "  if (a %% 2 == 1) {",
  "    return(FALSE)",
  "  }",
  "  if (!(a %% 2 == 1)) {",
  "    return(TRUE)",
  "  }",
  "  if (!(a %% 2 == 1)) {",
  "    print('Yaay')",
  "  }",
  "}",
  "is_even(150)",
  sep = "\n"
)
cat(code)
opt_code <- opt_cond_thread(list(code))
cat(opt_code$codes[[1]])
bmark_res <- microbenchmark({
  eval(parse(text = code))
}, {
  eval(parse(text = opt_code))
})
autoplot(bmark_res)
speed_up(bmark_res)
```